### PR TITLE
avoid division by zero in ceres::sqrt(0)

### PR DESCRIPTION
**Pull request by evenex**
implementation of ceres::sqrt divides by its argument, add checks to avoid this.
Also disable deprecated-copy warning, as latest Eigen3 seems to trigger -Werror=deprecated-copy

### CODE REVIEW

**Comment by NikolausDemmel** [comment]
Wouldn't this wrongly return a finite derivate for the theta output variable (assuming the `Scalar` template argument is also Jet) in the theta=0 case?

Why is the division by 0 an issue? Does it not just return inf / nan for the derivatives (which is correct, since the sqrt is not differentiable at 0), or does it actually crash?

Related: https://github.com/ceres-solver/ceres-solver/issues/447#issuecomment-527855933

File: sophus/so3.hpp [hunk 1]
```diff
@@ -585,7 +585,7 @@ class SO3 : public SO3Base<SO3<Scalar_, Options>> {
     using std::sin;
     using std::sqrt;
     Scalar theta_sq = omega.squaredNorm();
-    *theta = sqrt(theta_sq);
```
Comment (strasdat) [inline sophus/so3.hpp]: This I believe we should leave as is, and hence the derivative will of sqrt(x) at x=0 will be inf - following @NikolausDemmel's argument.

Comment (evenex) [reply sophus/so3.hpp]: n is also used on line 281

File: sophus/so3.hpp [hunk 2]
```diff
@@ -245,7 +245,7 @@ class SO3Base {
     using std::atan;
     using std::sqrt;
     Scalar squared_n = unit_quaternion().vec().squaredNorm();
-    Scalar n = sqrt(squared_n);
+    Scalar n = (squared_n == Scalar(0))? squared_n : sqrt(squared_n);
```
[the remark this answers is not in the document]
Comment (evenex) [reply sophus/so3.hpp]: n is also used on line 281

**Pull request closed without merging**